### PR TITLE
Refactor tunnel code

### PR DESCRIPTION
**Pull request by NielsZeilemaker**
Overal cleanup of the code. Biggest changed:
* Removed retrying for circuits
* Fixed Extended message (the created message was actually forwarded back to the origin, instead of the last hop). Causing the Extended message never to be created :smile: 
* Tried to simply the crypto stuffs, but failed
* Removed the tunnel_to_end methods

For now I left on all the debug logging. Laurens, have a look at the number of (very small) packets being sent to the same ip/ports during a libtorrent download (e.g. the test_hidden_community). I think adding a buffer, and batch sending those will substantially improve download speeds.

### CODE REVIEW

**Comment by NielsZeilemaker** [comment]
retest this please

File: Tribler/community/tunnel/routing.py [hunk 1]
```diff
@@ -95,7 +96,14 @@ def tunnel_data(self, destination, payload):
         @return bool: whether the tunnel request has succeeded, this is in no
          way an acknowledgement of delivery!
         """
-        return self.proxy.tunnel_data_to_end(destination, payload, self)
+
+        self._logger.error("Tunnel data (len %d) to end for circuit %s with ultimate destination %s", len(payload),
```
Comment (whirm) [inline Tribler/community/tunnel/routing.py]: We should lower this, any level will be shown if a test fail and if you are debugging the tunnels locally, you can always edit logger.conf to lower this particular logger.

Comment (NielsZeilemaker) [inline Tribler/community/tunnel/routing.py]: I know, I left it at error for now to show @lfdversluis know that there are a huge amount of small packets being sent.

File: Tribler/community/tunnel/tunnel_community.py [hunk 2]
```diff
@@ -458,15 +472,12 @@ def do_remove(self):
...
+    def create_circuit(self, goal_hops, ctype=CIRCUIT_TYPE_DATA, callback=None, required_endpoint=None, info_hash=None):
         assert required_endpoint is None or isinstance(required_endpoint, tuple), type(required_endpoint)
         assert required_endpoint is None or len(required_endpoint) == 3, required_endpoint
-        retry_lambda = first_hop = None
+        first_hop = None
 
-        if max_retries > 0:
-            retry_lambda = lambda h = goal_hops, t = ctype, c = callback, r = max_retries - 1, e = required_endpoint: \
-                self.create_circuit(h, t, c, r, e)
+        self.tunnel_logger.warning("Creating a new circuit of length %d", goal_hops)
```
Comment (whirm) [inline Tribler/community/tunnel/tunnel_community.py]: This one too

File: Tribler/community/tunnel/tunnel_community.py [hunk 3]
```diff
@@ -739,53 +750,92 @@ def check_create(self, messages):
...
+                continue
+            yield message
+
+    def check_destroy(self, messages):
+        for message in messages:
+            circuit_id = message.payload.circuit_id
+            cand_sock_addr = message.candidate.sock_addr
+
+            if circuit_id in self.relay_from_to:
+                pass
```
Comment (whirm) [inline Tribler/community/tunnel/tunnel_community.py]: Was that meant to be a `continue`?

Comment (NielsZeilemaker) [inline Tribler/community/tunnel/tunnel_community.py]: I copy/pasted that logic from the original on_destroy method. I was too afraid to touch it, so I simply converted into a pass :smile:

**Comment by lfdversluis** [comment]
I managed to run the test successfully once but didn't saw output in my terminal. Are the logs written to a file somewhere?

Also, I get segmentation faults when running the test:

``nosetests Tribler/Test/test_tunnel_community_positive.py``
Gives ``Segmentation fault (core dumped)``

**Comment by whirm** [comment]
@lfdversluis if a test passes, it's output is dropped by default. Have a look at nose's arguments if you want the output to be always printed.

If you get segfaults, run the tests through gdb so you can inspect the process state.

**Comment by whirm** [comment]
@lfdversluis I'm merging this so you can work from `devel`

**Pull request merged**